python/tk_multi_publish2/base_hooks/collector.py

```diff
@@ -155,10 +166,224 @@ def update_value(self):
+                                value_type, field_type))
+
+            # If this is coming from a linked property, we need to update
+            # the widget as well
+            if field_widget is not self._field_widgets[field_name]:
+                signals_blocked = self._field_widgets[field_name].blockSignals(True)
+                try:
+                    if field_value == self.MultiplesValue:
+                        self._field_widgets[field_name].set_value(self.MultiplesStr)
+                    elif field_value == self.NoneValue:
+                        self._field_widgets[field_name].set_value(self.NoneStr)
+                    else:
+                        self._field_widgets[field_name].set_value(field_value)
+                finally:
+                    self._field_widgets[field_name].blockSignals(signals_blocked)
+
+            # Update the fields dictionary with the new value
+            self._fields[field_name] = field_value
+
+            # Emit that our field value has changed
+            self.field_changed.emit()
+
+            # Recache the data
+            self.apply_changes()
+
+            # Update the ui
+            if do_full_refresh:
+                self.refresh_ui()
+
+        def gather_fields(self):
+            """
+            Gather the list of template fields
+            """
+            # Gather the fields for every input task
+            fields = {}
+            for item in self._items:
+                # Get the fields from the properties
+                if item.properties.get(self._name):
+                    for k, v in item.properties.get(self._name).iteritems():
+                        if k in fields:
+                            if fields[k] != v:
+                                fields[k] = self.MultiplesValue
+                        else:
+                            fields[k] = v
+
+                # Now pickup any overridden values already set via the UI
+                for field in fields.iterkeys():
+                    if field in self._fields:
+                        fields[field] = self._fields[field]
+
+            # Now update the member dict
+            self._fields = fields
+
+        def apply_changes(self):
+            """Store persistent data on the properties object"""
+            for item in self._items:
+                for key, value in self._fields.iteritems():
+                    if value == self.MultiplesValue or key in self._non_editable_fields:
+                        # Don't override value with multiples key,
+                        # or even keys that are not editable.
+                        continue
+                    # update the item.properties.fields
+                    item.properties[self._name][key] = value
+
+                for task in item._tasks:
+                    # re-init the task settings to pick up the new fields.
+                    task.init_task_settings()
```

Comment:
As I mentioned in the ticket, the call to `init_task_settings()` should not be triggered from within the `PropertyWidget` context.  The `PropertyWidget` should solely be responsible for interacting with its internal data structures and updating the properties for the item(s) its representing. All other data manipulation should be triggered using the `property_changed()` signal on the `PropertiesWidgetController`.  In our particular use case, we indeed want to call `init_task_settings()` if/when an item property is updated, but this is not necessarily the general case and hard coding it here 1) makes the incorrect assumption that this is the desired default behavior and 2) makes it difficult to extend/replace this behavior via inheritance.

Instead, in dialog.py, there should probably be something like:
```
widget = self._publish_manager._collector_instance.run_create_properties_widget(
            self.ui.property_settings_parent,
            items
        )
widget.property_changed.connect(lambda x: [self._publish_manager._collector_instance.run_on_properties_changed(item) for item in x.sender().items])
```

In collector_instance.py:
```
def run_on_properties_changed(self, item):
    with self._handle_plugin_error(None, "Error updating properties: %s"):
        self._hook_instance.on_properties_changed(self.settings, item)
```

In base_hooks/collector.py:
```
def on_properties_changed(self, settings, item):
    raise NotImplementedError
```

In hooks/collector.py:
```
def on_properties_changed(self, settings, item):
    for task in item.tasks:
        task.init_task_settings()
```

Reply:
Although, we are already `self.parent().setting_changed.connect(self.update_field)` of the widget to called `update_field`. Connecting the same signal twice doesn't work with signals, so we won't be able to use the above design.

[Qt Signals connected to multiple slots](https://stackoverflow.com/questions/3530590/qt-signals-and-slot-connected-twice-what-happens)

Reply:
`item.properties.fields` are only used to initialize task settings in the derived `hooks/publish.py` hook. It is not the base class default behavior, nor should it be. The purpose of any derived `PropertyWidget` is to update the selected item(s) property. Any additional processing can be triggered by this event, but should be handled externally.

Reply:
> will make us update the task_settings for any possible property update, even if they are not related to a task.
btw, that's what your current code is doing right now as well.

Reply:
I'm not really sure, if this works well. We don't know in what order the signal will call a slot, I couldn't find much documentation where it states that the slots are called in an order or can we even specify an order?

Without the `item.properties.fields` being updated i.e. `update_field` slot will need to be called before `on_properties_changed`, it won't make sense to call `init_task_settings` without the other.

Reply:
You need to simplify the signal processing. Properties are not interconnected like settings are, so there is no need to trigger an update on all other properties when one property is changed.

`update_value()` should be getting triggered by the `value_changed` signal from `self._value_widget` and `update_field()` by the `value_changed` signal from any of the field widgets. These, in turn, should emit their `value_changed()` signals, which will trigger the parent's `property_changed` signal, which will then trigger the code I have above.

You should be able to get the PropertyWidget from the `property_changed()` signal via `widget = self.sender().sender()`, which you can use to determine what property changed and then, since you have basically carte blanche to mutate how `on_properties_changed()` behaves, you can make it smarter to only call `task.init_task_settings()` for a specific subset of properties. If you feel like there isn't enough information getting passed via the signal, we can also discuss extending it to provide additional information.